src/components/v2v3/V2V3Project/V2V3ProjectSettings/pages/ProjectUpgradesPage/versions/useAvailableUpgrades.ts

```diff
@@ -0,0 +1,28 @@
+import { CV_V3 } from 'constants/cv'
+import { V2V3ContractsContext } from 'contexts/v2v3/Contracts/V2V3ContractsContext'
+import { V2V3ProjectContractsContext } from 'contexts/v2v3/ProjectContracts/V2V3ProjectContractsContext'
+import { useContext } from 'react'
+import { isEqualAddress } from 'utils/address'
+import { JBVersion } from './versions'
+
+export function useAvailableUpgrades(): JBVersion[] | undefined {
+  const { cvs, contracts } = useContext(V2V3ContractsContext)
+  const { contracts: projectContracts } = useContext(
+    V2V3ProjectContractsContext,
+  )
+
+  if (!cvs?.includes(CV_V3)) {
+    return ['3']
+  }
+
+  if (
+    projectContracts?.JBController &&
+    contracts?.JBController3_0_1 &&
+    !isEqualAddress(
+      projectContracts.JBController.address,
+      contracts.JBController3_0_1?.address,
+    )
+  ) {
+    return ['JBController3_0_1']
+  }
```

Comment:
**NOTE** for new upgrades we want to add, the dev will have to add a check to this hook.

Reply:
we could automate this a bit by doing something like:

```
const availableUpgrades: Record<JBVersion, ({projectContracts, cvs, ...}) => boolean> = {
  ...
}

const jbVersions: JBVersion[] = [...]

return jbVersions.filter(v => availableUpgrades({projectContracts, cvs}))
```

this way `availableUpgrades` would by type enforced whenever a new JBVersion is defined. but maybe not worth the trouble, may also scale awkwardly depending on future upgrades

Reply:
Yes absolutely love this suggestion! Will look into it.